src/ir.cpp

```diff
@@ -25194,12 +25194,40 @@ static IrInstGen *ir_analyze_instruction_cmpxchg(IrAnalyze *ira, IrInstSrcCmpxch
         return ira->codegen->invalid_inst_gen;
     }
 
+    ZigType *result_type = get_optional_type(ira->codegen, operand_type);
+
+    // special case zero bit types
+    if (type_has_one_possible_value(ira->codegen, operand_type) == OnePossibleValueYes) {
+        IrInstGen *result = ir_const(ira, &instruction->base.base, result_type);
+        set_optional_value_to_null(result->value);
+        return result;
+    }
+
     if (instr_is_comptime(casted_ptr) && casted_ptr->value->data.x_ptr.mut != ConstPtrMutRuntimeVar &&
         instr_is_comptime(casted_cmp_value) && instr_is_comptime(casted_new_value)) {
-        zig_panic("TODO compile-time execution of cmpxchg");
+        ZigValue *ptr_val = ir_resolve_const(ira, casted_ptr, UndefBad);
+        if (ptr_val == nullptr)
+            return ira->codegen->invalid_inst_gen;
+
+        ZigValue *op1_val = const_ptr_pointee(ira, ira->codegen, ptr_val, instruction->base.base.source_node);
+        if (op1_val == nullptr)
+            return ira->codegen->invalid_inst_gen;
+
+        ZigValue *op2_val = ir_resolve_const(ira, casted_cmp_value, UndefBad);
+        if (op2_val == nullptr)
+            return ira->codegen->invalid_inst_gen;
+
+        bool eql = const_values_equal(ira->codegen, op1_val, op2_val);
+        IrInstGen *result = ir_const(ira, &instruction->base.base, result_type);
+        if (eql) {
+            ir_analyze_store_ptr(ira, &instruction->base.base, casted_ptr, casted_new_value, false);
```

Comment:
You can use a simpler `copy_const_value` here.

Reply:
I can't seem get this to work with `copy_const_val`. This doesn't change the pointers value.
```c++

        bool eql = const_values_equal(ira->codegen, op1_val, op2_val);
        IrInstGen *result = ir_const(ira, &instruction->base.base, result_type);
        if (eql) {
            copy_const_val(ira->codegen, op1_val, op2_val);
            set_optional_value_to_null(result->value);
        } else {
            set_optional_payload(result->value, op1_val);
        }
        return result;

Reply:
`const_ptr_pointee` returns a pointer to the element pointed to so that should work.
Are you sure the logic is correct? You're doing `op1 <- op2` after checking that `op1 == op2`.

Reply:
I'm trying to make `ptr_val`, which is currently pointing at `op1_val`, to point to `op2_val`

Reply:
```zig
bool eql = const_values_equal(ira->codegen, op1_val, op2_val);
// snip...
        if (eql) {
            copy_const_val(ira->codegen, op1_val, op2_val);

Reply:
??

Reply:
Your code was a no-op, if `op1` and `op2` are equal the `copy_const_val` won't do anything.

Reply:
Ah, I confused `cmp_value` with `new_value`.